model/waterfall.go

```diff
@@ -13,54 +13,111 @@ import (
 )
 
 const (
+	DefaultWaterfallQueryCount   = 20
 	DefaultWaterfallVersionLimit = 5
 	MaxWaterfallVersionLimit     = 300
 )
 
 type WaterfallTask struct {
 	Id          string `bson:"_id" json:"_id"`
-	DisplayName string `bson:"display_name" json:"display_name,omitempty"`
-	Status      string `bson:"status" json:"status,omitempty"`
+	DisplayName string `bson:"display_name" json:"display_name"`
+	Status      string `bson:"status" json:"status"`
 }
 
 type WaterfallBuild struct {
 	Id          string          `bson:"_id" json:"_id"`
-	Activated   bool            `bson:"activated" json:"activated,omitempty"`
-	DisplayName string          `bson:"display_name" json:"display_name,omitempty"`
-	Version     string          `bson:"version" json:"version,omitempty"`
-	Tasks       []WaterfallTask `bson:"tasks" json:"tasks,omitempty"`
+	Activated   bool            `bson:"activated" json:"activated"`
+	DisplayName string          `bson:"display_name" json:"display_name"`
+	Version     string          `bson:"version" json:"version"`
+	Tasks       []WaterfallTask `bson:"tasks" json:"tasks"`
 }
 
 type WaterfallBuildVariant struct {
 	Id          string           `bson:"_id" json:"_id"`
-	DisplayName string           `bson:"display_name" json:"display_name,omitempty"`
-	Builds      []WaterfallBuild `bson:"builds" json:"builds,omitempty"`
+	DisplayName string           `bson:"display_name" json:"display_name"`
+	Builds      []WaterfallBuild `bson:"builds" json:"builds"`
 }
 
 type WaterfallOptions struct {
 	Limit      int      `bson:"-" json:"-"`
+	MaxOrder   int      `bson:"-" json:"-"`
+	MinOrder   int      `bson:"-" json:"-"`
 	Requesters []string `bson:"-" json:"-"`
 }
 
-// GetWaterfallVersions returns at most `opts.limit` versions for a given project.
-// TODO: It will eventually return `opts.limit` activated versions that also satisfy any given filters.
-func GetWaterfallVersions(ctx context.Context, projectId string, opts WaterfallOptions) ([]Version, error) {
+// GetActiveWaterfallVersions returns at most `opts.limit` activated versions for a given project.
```

Comment:
Might help to mention the sorting order since the results are always sorted but the direction can vary.

Also, is `opts.Limit` supposed to be used? It looks like it was removed.

Reply:
Good catch on `opts.Limit`, I re-added!